akka-docs/rst/java/code/docs/persistence/PersistenceQueryDocTest.java

```diff
@@ -44,48 +46,156 @@
 
   final ActorSystem system = ActorSystem.create();
   final ActorMaterializer mat = ActorMaterializer.create(system);
+  
+  static
+  //#advanced-journal-query-types
+  public class RichEvent {
+    public final Set<String >tags;
+    public final Object payload;
+    
+    public RichEvent(Set<String> tags, Object payload) {
+      this.tags = tags;
+      this.payload = payload;
+    }
+  }
+  //#advanced-journal-query-types
 
+  static
+  //#advanced-journal-query-types
+  // a plugin can provide:
+  public final class QueryMetadata{
+    public final boolean deterministicOrder;
+    public final boolean infinite;
+
+    public QueryMetadata(boolean deterministicOrder, boolean infinite) {
+      this.deterministicOrder = deterministicOrder;
+      this.infinite = infinite;
+    }
+  }
+  //#advanced-journal-query-types
+
+  static
   //#my-read-journal
-  class MyReadJournal implements ReadJournal {
-    private final ExtendedActorSystem system;
+  public class MyReadJournalProvider implements ReadJournalProvider {
+    private final MyJavadslReadJournal javadslReadJournal;
 
-    public MyReadJournal(ExtendedActorSystem system, Config config) {
-      this.system = system;
+    public MyReadJournalProvider(ExtendedActorSystem system, Config config) {
+      this.javadslReadJournal = new MyJavadslReadJournal(system, config);
+    }
+    
+    @Override
+    public MyScaladslReadJournal scaladslReadJournal() {
+      return new MyScaladslReadJournal(javadslReadJournal);
     }
 
-    final FiniteDuration defaultRefreshInterval = FiniteDuration.create(3, TimeUnit.SECONDS);
-
-    @SuppressWarnings("unchecked")
-    public <T, M> Source<T, M> query(Query<T, M> q, Hint... hints) {
-      if (q instanceof EventsByTag) {
-        final EventsByTag eventsByTag = (EventsByTag) q;
-        final String tag = eventsByTag.tag();
-        long offset = eventsByTag.offset();
-
-        final Props props = MyEventsByTagPublisher.props(tag, offset, refreshInterval(hints));
-
-        return (Source<T, M>) Source.<EventEnvelope>actorPublisher(props)
-                                    .mapMaterializedValue(noMaterializedValue());
-      } else {
-        // unsuported
-        return Source.<T>failed(
-          new UnsupportedOperationException(
-            "Query " + q + " not supported by " + getClass().getName()))
-                     .mapMaterializedValue(noMaterializedValue());
-      }
+    @Override
+    public MyJavadslReadJournal javadslReadJournal() {
```

Comment:
ok, I don't mind `readJournal` and `getReadJournal` since that is the corresponding user api. hmm